snuba/datasets/errors_replacer.py

```diff
@@ -373,17 +375,30 @@ def process_delete_groups(
 
 
 def process_tombstone_events(
-    message: Mapping[str, Any], required_columns: Sequence[str]
+    message: Mapping[str, Any],
+    required_columns: Sequence[str],
+    state_name: ReplacerState,
 ) -> Optional[Replacement]:
     event_ids = message["event_ids"]
     if not event_ids:
         return None
 
+    old_primary_hash = message.get("old_primary_hash")
```

Comment:
Under what circumstances would a `tombstone_event` message not have this key? Shouldn't it always be present?

Reply:
when the event genuinely should be deleted and not because of errors table PK

we don't expose "event deletion" as standalone feature but we do expose the option to reprocess half of an issue and delete the other half. the way we implement this is by reinserting some events, tombstoning others, and then sending `exclude_groups` message at the end to delete the old group id

Reply:
I see. Let me confirm what happens in the 4 scenarios now-

1. events table with old primary hash - nothing happens, ClickHouse merging will eventually take care of it
2. events table without old primary hash - ignore primary_hash column, write deleted = 1 in any rows matching event id
3. errors table with old primary hash - we write deleted = 1 only in the rows also matching that old primary hash
4. errors table without old primary hash - ignore primary_hash column, write deleted = 1 in any rows matching event id

Is that right?

Reply:
yup that sounds right

Reply:
Sorry, I think i'm a bit confused about some details still. Is the process_tombstone event also accompanied by a reinsertion event?

If so, why do we need to do any ClickHouse query here for cases 2 and 4 at all? Since the primary key will be the same, won't ClickHouse just merge those rows just fine like in scenario 1 without us ever writing anything in `deleted`?

Reply:
from sentry POV there is really one usecase: we want to reinsert the event with a different primary_hash. In the case of events table this is a noop (2), and in the case of errors table the PK changes so the old PK has to be tombstoned.

Sentry does not know the difference between the tables. Sentry emits a tombstone_events message with `old_primary_hash` in the case it is about to reinsert the same event ID with a different primary_hash. Snuba then decides whether that is supposed to result in an actual tombstone, based on which table is used.

>If so, why do we need to do any ClickHouse query here for cases 2 and 4 at all?

In cases 2 and 4, there will not be any reinsert after the tombstone. Those events really are supposed to be gone for good.

Let's say you reprocess an issue with two events. There is a dialogue where the user can choose to only reprocess one event A and delete the other event B. This results in the following operations to Snuba:

* `tombstone_events(B)` with `old_primary_hash=null` -- from sentry POV this can result in either case 2 or 4, but really the expectation is just that the event can't be queried anymore
* `tombstone_events(A)` with `old_primary_hash=123abc` *in case the new primary_hash is different.* -- from sentry POV this can result in either case 1 or 3, but really the expectation is just that the event won't appear twice after reinsert
* `eventstream.insert(A2)` (same event ID as A)
* `exclude_groups`

Reply:
Makes sense, did not realise we were also allowing individual events to be targeted for deletion without reinsertion.

Reply:
I think it'd be good to have a test to cover cases 2 and 4, from what I can tell only 1 and 3 are covered

Reply:
case 4 is covered via pytest parametrization, I now added a case for 2